File: related_scripts/verification/step_E/benchmark_gridding.py
"""Benchmark: Python Sibson gridding speed at production scale.

Tests:
1. Weight matrix computation (sibson_weight_matrix)
2. Batched sparse matmul (W @ V)
3. Full gridding pipeline

Pair with benchmark_gridding.m for MATLAB comparison.
"""
import sys
import time
from pathlib import Path

import numpy as np

sys.path.insert(0, str(Path(__file__).resolve().parents[3] / "src"))
from caz.gridding import GridSpec, _albers_forward
from caz.natneighbor import sibson_weight_matrix

SPEC = GridSpec(lat_min=15.0, lat_max=75.0, lon_min=-170.0, lon_max=-50.0, res_deg=0.20)


def setup_grid():
    glat, glon = SPEC.lat_lon()
    ny, nx = len(glat), len(glon)
    longrid, latgrid = np.meshgrid(glon, glat)
    x_grd, y_grd = _albers_forward(longrid.ravel(), latgrid.ravel())
    qpts = np.column_stack([x_grd, y_grd])
    return qpts, ny, nx


def make_stations(n_st, seed=42):
    np.random.seed(seed)
    lat = np.random.uniform(20, 70, n_st).astype(np.float32)
    lon = np.random.uniform(-165, -55, n_st).astype(np.float32)
    x, y = _albers_forward(lon, lat)
    return np.column_stack([x, y])


def main():
    print("=== Gridding Benchmark (Python, Numba-parallel) ===\n")
    qpts, ny, nx = setup_grid()
    M = len(qpts)
    print(f"Grid: {ny}x{nx} = {M} query points\n")

    # Warmup
    small = make_stations(20)
    _ = sibson_weight_matrix(small, qpts[:100])

    results = {}
    for label, n_st in [("Small (800)", 800), ("Production (7000)", 7000)]:
        print(f"--- {label} stations ---")
        st = make_stations(n_st)

        t0 = time.perf_counter()
        W = sibson_weight_matrix(st, qpts)
        t_wm = time.perf_counter() - t0
        print(f"  Weight matrix: {t_wm:.3f}s (nnz={W.nnz})")

        V = np.random.randn(n_st, 125)
        t0 = time.perf_counter()
        G = W @ V
        t_mm = time.perf_counter() - t0
        print(f"  Batched matmul (125 cols): {t_mm:.3f}s")

        vals = np.random.randn(n_st)
        t0 = time.perf_counter()
        for _ in range(125):
            r = W @ vals
        t_mv = time.perf_counter() - t0
        print(f"  125x single matmul: {t_mv:.3f}s")

        results[label] = {"weight_matrix": t_wm, "batch_mm": t_mm, "single_mv": t_mv}
        print()

    # Summary
    print("=== Summary ===")
    for label, r in results.items():
        print(f"  {label}:")
        for k, v in r.items():
            print(f"    {k}: {v:.3f}s")

    # Production timing from run_varcorr_cal.py
    import json
    for var in ["tas", "pr"]:
        p = Path(r"D:\Dataset\DPastCliM-NA\interim\grid_cal") / f"timing_varcorr_{var}.json"
        if p.exists():
            with open(p) as f:
                t = json.load(f)
            print(f"\n  Production {var} (run_varcorr_cal.py):")
            for k, v in t.items():
                print(f"    {k}: {v:.1f}s")

    print("\nDone.")


if __name__ == "__main__":
    main()

File: related_scripts/verification/step_E/benchmark_gridding.m
% Benchmark MATLAB scatteredInterpolant gridding speed.
% Pair with benchmark_gridding.py for Python comparison.

fprintf('=== Gridding Benchmark (MATLAB scatteredInterpolant) ===\n\n');

% Use raw coordinates (no projection) since Mapping Toolbox unavailable.
% Timing comparison is still valid: same grid size and interpolation method.
lat_range = [15 75];
lon_range = [-170 -50];
res = 0.20;

lat = lat_range(1):res:lat_range(2);
lon = lon_range(1):res:lon_range(2);
ny = length(lat);
nx = length(lon);
[longrid, latgrid] = meshgrid(lon, lat);
xgrid = longrid(:);
ygrid = latgrid(:);
M = length(xgrid);
fprintf('Grid: %dx%d = %d query points\n\n', ny, nx, M);

rng(42);
sizes = [800, 7000];
for si = 1:length(sizes)
    n_st = sizes(si);
    x_st = -165 + 110*rand(n_st, 1);
    y_st = 20 + 50*rand(n_st, 1);
    vals = randn(n_st, 1);

    fprintf('--- %d stations ---\n', n_st);

    % Single interpolation (scatteredInterpolant creation + evaluation)
    tic;
    F = scatteredInterpolant(x_st, y_st, vals, 'natural', 'nearest');
    result = F(xgrid, ygrid);
    t_single = toc;
    fprintf('  Single interp (create+eval): %.3f s\n', t_single);

    % 125 interpolations (same stations, different values)
    V = randn(n_st, 125);
    tic;
    for k = 1:125
        F.Values = V(:,k);
        result = F(xgrid, ygrid);
    end
    t_batch = toc;
    fprintf('  125x interp (reuse F): %.3f s\n', t_batch);
    fprintf('  Per-timestep: %.3f s\n\n', t_batch/125);
end

fprintf('Done.\n');
